File: src/frontend/nfe/lib/app/data/model/nfe_detalhe_imposto_issqn_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeDetalheImpostoIssqnModel {
	int? id;
	int? idNfeDetalhe;
	double? baseCalculoIssqn;
	double? aliquotaIssqn;
	double? valorIssqn;
	int? municipioIssqn;
	int? itemListaServicos;
	double? valorDeducao;
	double? valorOutrasRetencoes;
	double? valorDescontoIncondicionado;
	double? valorDescontoCondicionado;
	double? valorRetencaoIss;
	String? indicadorExigibilidadeIss;
	String? codigoServico;
	int? municipioIncidencia;
	int? paisSevicoPrestado;
	String? numeroProcesso;
	String? indicadorIncentivoFiscal;

	NfeDetalheImpostoIssqnModel({
		this.id,
		this.idNfeDetalhe,
		this.baseCalculoIssqn,
		this.aliquotaIssqn,
		this.valorIssqn,
		this.municipioIssqn,
		this.itemListaServicos,
		this.valorDeducao,
		this.valorOutrasRetencoes,
		this.valorDescontoIncondicionado,
		this.valorDescontoCondicionado,
		this.valorRetencaoIss,
		this.indicadorExigibilidadeIss,
		this.codigoServico,
		this.municipioIncidencia,
		this.paisSevicoPrestado,
		this.numeroProcesso,
		this.indicadorIncentivoFiscal,
	});

	static List<String> dbColumns = <String>[
		'id',
		'base_calculo_issqn',
		'aliquota_issqn',
		'valor_issqn',
		'municipio_issqn',
		'item_lista_servicos',
		'valor_deducao',
		'valor_outras_retencoes',
		'valor_desconto_incondicionado',
		'valor_desconto_condicionado',
		'valor_retencao_iss',
		'indicador_exigibilidade_iss',
		'codigo_servico',
		'municipio_incidencia',
		'pais_sevico_prestado',
		'numero_processo',
		'indicador_incentivo_fiscal',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Base Calculo Issqn',
		'Aliquota Issqn',
		'Valor Issqn',
		'Municipio Issqn',
		'Item Lista Servicos',
		'Valor Deducao',
		'Valor Outras Retencoes',
		'Valor Desconto Incondicionado',
		'Valor Desconto Condicionado',
		'Valor Retencao Iss',
		'Indicador Exigibilidade Iss',
		'Codigo Servico',
		'Municipio Incidencia',
		'Pais Sevico Prestado',
		'Numero Processo',
		'Indicador Incentivo Fiscal',
	];

	NfeDetalheImpostoIssqnModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeDetalhe = jsonData['idNfeDetalhe'];
		baseCalculoIssqn = jsonData['baseCalculoIssqn']?.toDouble();
		aliquotaIssqn = jsonData['aliquotaIssqn']?.toDouble();
		valorIssqn = jsonData['valorIssqn']?.toDouble();
		municipioIssqn = jsonData['municipioIssqn'];
		itemListaServicos = jsonData['itemListaServicos'];
		valorDeducao = jsonData['valorDeducao']?.toDouble();
		valorOutrasRetencoes = jsonData['valorOutrasRetencoes']?.toDouble();
		valorDescontoIncondicionado = jsonData['valorDescontoIncondicionado']?.toDouble();
		valorDescontoCondicionado = jsonData['valorDescontoCondicionado']?.toDouble();
		valorRetencaoIss = jsonData['valorRetencaoIss']?.toDouble();
		indicadorExigibilidadeIss = NfeDetalheImpostoIssqnDomain.getIndicadorExigibilidadeIss(jsonData['indicadorExigibilidadeIss']);
		codigoServico = jsonData['codigoServico'];
		municipioIncidencia = jsonData['municipioIncidencia'];
		paisSevicoPrestado = jsonData['paisSevicoPrestado'];
		numeroProcesso = jsonData['numeroProcesso'];
		indicadorIncentivoFiscal = NfeDetalheImpostoIssqnDomain.getIndicadorIncentivoFiscal(jsonData['indicadorIncentivoFiscal']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeDetalhe'] = idNfeDetalhe != 0 ? idNfeDetalhe : null;
		jsonData['baseCalculoIssqn'] = baseCalculoIssqn;
		jsonData['aliquotaIssqn'] = aliquotaIssqn;
		jsonData['valorIssqn'] = valorIssqn;
		jsonData['municipioIssqn'] = municipioIssqn;
		jsonData['itemListaServicos'] = itemListaServicos;
		jsonData['valorDeducao'] = valorDeducao;
		jsonData['valorOutrasRetencoes'] = valorOutrasRetencoes;
		jsonData['valorDescontoIncondicionado'] = valorDescontoIncondicionado;
		jsonData['valorDescontoCondicionado'] = valorDescontoCondicionado;
		jsonData['valorRetencaoIss'] = valorRetencaoIss;
		jsonData['indicadorExigibilidadeIss'] = NfeDetalheImpostoIssqnDomain.setIndicadorExigibilidadeIss(indicadorExigibilidadeIss);
		jsonData['codigoServico'] = codigoServico;
		jsonData['municipioIncidencia'] = municipioIncidencia;
		jsonData['paisSevicoPrestado'] = paisSevicoPrestado;
		jsonData['numeroProcesso'] = numeroProcesso;
		jsonData['indicadorIncentivoFiscal'] = NfeDetalheImpostoIssqnDomain.setIndicadorIncentivoFiscal(indicadorIncentivoFiscal);
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeDetalhe = plutoRow.cells['idNfeDetalhe']?.value;
		baseCalculoIssqn = plutoRow.cells['baseCalculoIssqn']?.value?.toDouble();
		aliquotaIssqn = plutoRow.cells['aliquotaIssqn']?.value?.toDouble();
		valorIssqn = plutoRow.cells['valorIssqn']?.value?.toDouble();
		municipioIssqn = plutoRow.cells['municipioIssqn']?.value;
		itemListaServicos = plutoRow.cells['itemListaServicos']?.value;
		valorDeducao = plutoRow.cells['valorDeducao']?.value?.toDouble();
		valorOutrasRetencoes = plutoRow.cells['valorOutrasRetencoes']?.value?.toDouble();
		valorDescontoIncondicionado = plutoRow.cells['valorDescontoIncondicionado']?.value?.toDouble();
		valorDescontoCondicionado = plutoRow.cells['valorDescontoCondicionado']?.value?.toDouble();
		valorRetencaoIss = plutoRow.cells['valorRetencaoIss']?.value?.toDouble();
		indicadorExigibilidadeIss = plutoRow.cells['indicadorExigibilidadeIss']?.value != '' ? plutoRow.cells['indicadorExigibilidadeIss']?.value : 'AAA';
		codigoServico = plutoRow.cells['codigoServico']?.value;
		municipioIncidencia = plutoRow.cells['municipioIncidencia']?.value;
		paisSevicoPrestado = plutoRow.cells['paisSevicoPrestado']?.value;
		numeroProcesso = plutoRow.cells['numeroProcesso']?.value;
		indicadorIncentivoFiscal = plutoRow.cells['indicadorIncentivoFiscal']?.value != '' ? plutoRow.cells['indicadorIncentivoFiscal']?.value : 'AAA';
	}	

	NfeDetalheImpostoIssqnModel clone() {
		return NfeDetalheImpostoIssqnModel(
			id: id,
			idNfeDetalhe: idNfeDetalhe,
			baseCalculoIssqn: baseCalculoIssqn,
			aliquotaIssqn: aliquotaIssqn,
			valorIssqn: valorIssqn,
			municipioIssqn: municipioIssqn,
			itemListaServicos: itemListaServicos,
			valorDeducao: valorDeducao,
			valorOutrasRetencoes: valorOutrasRetencoes,
			valorDescontoIncondicionado: valorDescontoIncondicionado,
			valorDescontoCondicionado: valorDescontoCondicionado,
			valorRetencaoIss: valorRetencaoIss,
			indicadorExigibilidadeIss: indicadorExigibilidadeIss,
			codigoServico: codigoServico,
			municipioIncidencia: municipioIncidencia,
			paisSevicoPrestado: paisSevicoPrestado,
			numeroProcesso: numeroProcesso,
			indicadorIncentivoFiscal: indicadorIncentivoFiscal,
		);			
	}

	
}

File: backend/nfe/src/model/nfe_detalhe_imposto_issqn_model.py
from src import db


class NfeDetalheImpostoIssqnModel(db.Model):
    __tablename__ = 'nfe_detalhe_imposto_issqn'

    id = db.Column(db.Integer, primary_key=True)
    base_calculo_issqn = db.Column(db.Float)
    aliquota_issqn = db.Column(db.Float)
    valor_issqn = db.Column(db.Float)
    municipio_issqn = db.Column(db.Integer)
    item_lista_servicos = db.Column(db.Integer)
    valor_deducao = db.Column(db.Float)
    valor_outras_retencoes = db.Column(db.Float)
    valor_desconto_incondicionado = db.Column(db.Float)
    valor_desconto_condicionado = db.Column(db.Float)
    valor_retencao_iss = db.Column(db.Float)
    indicador_exigibilidade_iss = db.Column(db.String(1))
    codigo_servico = db.Column(db.String(20))
    municipio_incidencia = db.Column(db.Integer)
    pais_sevico_prestado = db.Column(db.Integer)
    numero_processo = db.Column(db.String(30))
    indicador_incentivo_fiscal = db.Column(db.String(1))
    id_nfe_detalhe = db.Column(db.Integer, db.ForeignKey('nfe_detalhe.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_detalhe = data.get('idNfeDetalhe')
        self.base_calculo_issqn = data.get('baseCalculoIssqn')
        self.aliquota_issqn = data.get('aliquotaIssqn')
        self.valor_issqn = data.get('valorIssqn')
        self.municipio_issqn = data.get('municipioIssqn')
        self.item_lista_servicos = data.get('itemListaServicos')
        self.valor_deducao = data.get('valorDeducao')
        self.valor_outras_retencoes = data.get('valorOutrasRetencoes')
        self.valor_desconto_incondicionado = data.get('valorDescontoIncondicionado')
        self.valor_desconto_condicionado = data.get('valorDescontoCondicionado')
        self.valor_retencao_iss = data.get('valorRetencaoIss')
        self.indicador_exigibilidade_iss = data.get('indicadorExigibilidadeIss')
        self.codigo_servico = data.get('codigoServico')
        self.municipio_incidencia = data.get('municipioIncidencia')
        self.pais_sevico_prestado = data.get('paisSevicoPrestado')
        self.numero_processo = data.get('numeroProcesso')
        self.indicador_incentivo_fiscal = data.get('indicadorIncentivoFiscal')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeDetalhe': self.id_nfe_detalhe,
            'baseCalculoIssqn': self.base_calculo_issqn,
            'aliquotaIssqn': self.aliquota_issqn,
            'valorIssqn': self.valor_issqn,
            'municipioIssqn': self.municipio_issqn,
            'itemListaServicos': self.item_lista_servicos,
            'valorDeducao': self.valor_deducao,
            'valorOutrasRetencoes': self.valor_outras_retencoes,
            'valorDescontoIncondicionado': self.valor_desconto_incondicionado,
            'valorDescontoCondicionado': self.valor_desconto_condicionado,
            'valorRetencaoIss': self.valor_retencao_iss,
            'indicadorExigibilidadeIss': self.indicador_exigibilidade_iss,
            'codigoServico': self.codigo_servico,
            'municipioIncidencia': self.municipio_incidencia,
            'paisSevicoPrestado': self.pais_sevico_prestado,
            'numeroProcesso': self.numero_processo,
            'indicadorIncentivoFiscal': self.indicador_incentivo_fiscal,
        }